[dynamo] Setup the globals for guard_fn without a reference to f_locals

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118448
* __->__ #118447
* #118420
* #118975

UPDATE - I changed the PR because from discussion with @jansel it was clear that someone else was holding on to a reference to f_locals. This PR now solves that problem first. I removed the eval_frame.c part because it was failing tests that use `exec` or `eval` with weird error like `no no locals found when storing 'math'`. I would debug that in a separate PR.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

anijain2305 committed:
[dynamo][eval_frame] Move f_locals clearing to eval_frame.c

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -1018,6 +1018,17 @@ static PyObject* _custom_eval_frame(
+    // test_release_input_memory). Therefore, we clear the dictionary here. The
+    // later call to eval_custom_code uses the fast representation for accessing
+    // the locals.
+    PyDict_Clear(frame->f_locals);
```
I don't think this is right.  My understanding is:
1) `frame->f_locals` starts as `NULL`
2) `PyFrame_FastToLocalsWithError` changes it to a `dict`
3) Here we should change it back to `NULL`

Something like:
```
Py_XDECREF(frame->f_locals);
frame->f_locals = NULL;
```

anijain2305 replied:
I checked and yes `frame->f_locals` starts as `NULL`. So, your solution makes sense.

But this test still fails
https://github.com/pytorch/pytorch/blob/2ed0af2bdeb2a5e0df78b22eab04d468f8db239e/test/dynamo/test_misc.py#L4067-L4080

Which likely means someone else is also holding on to f_locals or its contents in some way.

jansel replied:
Is there a pointer sticking around in the compiler code somewhere?

Perhaps use, `gc.get_referrers()` to find them,
https://docs.python.org/3/library/gc.html#gc.get_referrers

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Move f_locals clearing to eval_frame.c"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Setup the globals for guard_fn without a reference to f_locals"


UPDATE - I changed the PR because from discussion with jansel it was clear that someone else was holding on to a reference to f_locals. This PR now solves that problem first. I removed the eval_frame.c part because it was failing tests that use `exec` or `eval` with weird error like `no no locals found when storing 'math'`. I would debug that in a separate PR.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented:
@jansel This is ready for review.

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge
